### ISSUE TITLE
After a cluster upgrade, replication seems to work but Orchestrator fails to update cluster topology (connection refused)

### ISSUE THREAD

**Issue by vitobotta** [presslabs/mysql-operator#446]
Hi! I am using DigitalOcean's managed kubernetes and yesterday I replaced a node pool with another, which caused one node per time to be replaced. When the current MySQL master was evicted from a node being replaced, MySQL seemed to failover to the slave, but since then Orchestrator is unable to show the topology in the UI - I did see a message once saying that the cluster had been recovered or something like that, but now all I see is this:

<img width="385" alt="Screenshot 2020-01-07 at 11 26 44" src="https://user-images.githubusercontent.com/283811/71885967-0df63d80-3144-11ea-8b98-39191c0f9344.png">

I exec'ed into the orchestrator container and if I run something like

```
orchestrator -c topology -i mysql-cluster-mysql-0.mysql
```

All is get is this:

```
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 CRITICAL dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR ContinuousRegistration: RegisterNode failed: dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 CRITICAL dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 CRITICAL dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 ERROR dial tcp :3306: connect: connection refused
2020-01-07 09:46:33 FATAL dial tcp :3306: connect: connection refused
```

The replication seems to be working because when I save data the slave has the updated data. However I am not sure of what's going on with the Orchestrator and if there still is a problem. Now it's just me using the cluster but I am worried that these issues might happen regularly once I am in production, because DigitalOcean frequently replaces nodes for security updates etc, besides Kubernetes upgrades. 

What can I do to check if my cluster is actually OK and what can I do to prevent issues when cluster updates occur? 

I'm not familiar with Orchestrator so forgive me if I am missing something obvious.

Thanks in advance!

**Comment by vitobotta** [comment]
In the master's log I see 

```
Access denied for user ''@'10.244.2.58' (using password: NO)
```

And that is the IP of the orchestrator. It's trying to login with no username?

**Comment by vitobotta** [comment]
I don't understand, I tried several times restoring from a backup taken before the cluster upgrade and again, the replication seems to work but Orchestrator shows the same thing as shown in the image above plus this:

<img width="331" alt="Screenshot 2020-01-07 at 14 19 17" src="https://user-images.githubusercontent.com/283811/71895081-bb734c00-3158-11ea-9ce0-bb8c70f3274e.png">

I deleted the cluster, uninstalled the operator, deleted the namespace and restored from the backup. Always the same. Is there a problem with the current version? I have 0.3.5 now. @AMecea

**Comment by AMecea** [comment]
Hi @vitobotta, please show me the cluster status (`kubectl describe mysql <mysql-cluster-name>`).

Did you set a password for orchestrator topology? Check the secret `<release-name>-mysql-operator-orc` to have the right credentials set.

**Comment by vitobotta** [comment]
Hi @AMecea ! This is the describe: https://pastebin.com/raw/R5GF5pSY

I didn't set any password.  I restored again and get the same thing but with a lag of 52K seconds or so. I don't understand what happened, never had an issue before

**Comment by vitobotta** [comment]
After restoring the master svc doesn't have any endpoints either

**Comment by vitobotta** [comment]
Interesting... I restored as a single replica then I tried "Reset slave" on that replica. Orchestrator immediately showed the replica as "Master" instead of "Co master". Then I added another replica by editing the cluster and it's replicating from the master and Orchestrator seems OK now. 

What happened? Why is the main replica as co master after restoring from backup? I am now going to delete and try restoring a more recent backup from after the cluster upgrade, to see if that works too.

**Comment by vitobotta** [comment]
Yeah that trick seems to have worked with a more recent backup as well. I would appreciate an explanation of how this stuff works and what's the expected behaviour when cluster upgrade etc occur, since I am pretty new to this stuff. Thanks! :)

**Comment by vitobotta** [comment]
Shit.. I spoke too soon :p The master service has no endpoints after that....

**Comment by vitobotta** [comment]
So if I remove a pod and let recreate it, the endpoints get updated it seems... 

However: I first restored the cluster with 2 replicas, then I tried to add a third replica but that one wouldn't start properly, complaining about duplicate primary keys. I tried restoring with three replicas from the beginning and I see this in the third replica:

```
2020-01-07T14:08:45.442330Z 98 [Note] 'CHANGE MASTER TO FOR CHANNEL '' executed'. Previous state master_host='mysql-cluster-mysql-0.mysql.mysql', master_port= 3306, master_log_file='', master_log_pos= 4, master_bind=''. New state master_host='mysql-cluster-mysql-0.mysql.mysql', master_port= 3306, master_log_file='', master_log_pos= 4, master_bind=''.
2020-01-07T14:08:46.584537Z 98 [Note] Aborted connection 98 to db: 'unconnected' user: 'sys_operator' host: '10.244.3.229' (Got an error reading communication packets)
```

I have been using this operator for a few months and haven't had problems with it before. Now I am having problems with replication, orchestrator and restoring from backups properly....

When I install the operator with Helm it installs 0.3.5. Which version should I be using? The Releases page here on Github shows that there are tags for 0.3.2-5, but the last "release" is 0.3.1. Should I be using that one instead?

**Comment by vitobotta** [comment]
I am basically unable to restore any backup.... I am trying with several backups from before the cluster upgrade and nothing works. This last attempt I've restored a backup from two days ago with a single replica. It's up and running / ready but no endpoints in the services and Orchestrator shows the single replica as "Co master" with a long lag and in the details panel I see this:

<img width="827" alt="Screenshot 2020-01-07 at 16 40 09" src="https://user-images.githubusercontent.com/283811/71903299-72c58e00-316c-11ea-94f6-bd1640b277e7.png">

Why is mysql-cluster-mysql-0.mysql.mysql:3306 set as master for itself? 

@AMecea I am really sorry to bother you.... but I don't know what else to try :(

**Comment by vitobotta** [comment]
There's something wrong with my backups. I tried creating a brand new cluster instead of restoring and all looks good.

**Comment by AMecea** [comment]
I've never encountered this issue before, I will do more debugging. Version v0.3.5 should not be more different than v0.3.1 and I don't think that it's the root cause for this.

Let's move this converstation on slack.

**Comment by AMecea** [comment]
fixed by #447
